### PR TITLE
New Resource: `azurerm_system_center_virtual_machine_manager_cloud`

### PR DESCRIPTION
**Pull request by neil-yechenwei**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR is to support new resource System Center Virtual Machine Manager Cloud in SCVMM RP. Maybe these three PRs (https://github.com/hashicorp/terraform-provider-azurerm/pull/25449, https://github.com/hashicorp/terraform-provider-azurerm/pull/25451, https://github.com/hashicorp/terraform-provider-azurerm/pull/25429) need to be merged one by one.

## Change Log

* `azurerm_system_center_virtual_machine_manager_cloud` - support this new resource

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Test Result
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/basic (406.70s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/requiresImport (269.70s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/complete (326.31s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/update (592.14s)

## References
[Azure doc](https://learn.microsoft.com/en-us/azure/azure-arc/system-center-virtual-machine-manager/overview)
[REST API specs](https://github.com/Azure/azure-rest-api-specs/blob/fe9e014469660ac54057ceacba403c516575c5a1/specification/scvmm/resource-manager/Microsoft.ScVmm/stable/2023-10-07/scvmm.json#L467)

### CODE REVIEW

**Comment by rcskosir** [comment]
@neil-yechenwei :wave: I know this is currently a draft, but when it is ready for review can you please add in the new PR template? The extra information in the template helps with the review process and for traceability in the future for our users about what this change did.

**Comment by neil-yechenwei** [comment]
@rcskosir , thanks for the suggestion. I updated the PR template.

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (ms-zhenhua) [inline internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:85]: insensitively parsing should be used for API response only. Configured properties should be sensitively parsed.

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (ms-zhenhua) [inline internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:59]: could `system_center_virtual_machine_manager_server_inventory_item_id` and this resource reside in different resource groups? If not, `resource_group_name` is not necessary because it can be obtained from `system_center_virtual_machine_manager_server_inventory_item_id`?

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (ms-zhenhua) [inline internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:105]: please also update other places to keep consistent
```suggestion
					Type: pointer.To("customLocation"),
```

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (ms-zhenhua) [inline internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:153]:
```suggestion
			state := SystemCenterVirtualMachineManagerCloudModel{}
			state.Name = id.CloudName
			state.ResourceGroupName = id.ResourceGroupName
			if model := resp.Model; model != nil {	
				state.Location = location.Normalize(model.Location)
				state.CustomLocationId = pointer.From(model.ExtendedLocation.Name)
				state.SystemCenterVirtualMachineManagerServerInventoryItemId = pointer.From(model.Properties.InventoryItemId)
				state.Tags = pointer.From(model.Tags)
			}
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei, 
Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (neil-yechenwei) [reply internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:59]: They can be in different resource group.

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (neil-yechenwei) [reply internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:85]: Updated

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (neil-yechenwei) [reply internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:105]: Updated

File: internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go
Comment (neil-yechenwei) [reply internal/services/systemcentervirtualmachinemanager/system_center_virtual_machine_manager_cloud_resource.go:153]: Updated

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , thanks for the comments. I updated PR. Please take another look. Thanks.

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , thanks for the comments. I updated PR. Please take another look. Thanks.

**Review by ms-zhenhua — APPROVED** [review]
@neil-yechenwei , thank you for your updates. LGTM~

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (catriona-m) [inline website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: is it necessary to have this sleep for 1 minute? this may be an indication there's a bug in `azurerm_system_center_virtual_machine_manager_server`

**Review by catriona-m — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei thanks for opening this. I left one query about the test config, but otherwise this is looking good so far. Thanks!

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: It's required. Service team confirmed that it's not a bug since the backend operator creates CRD (Custom Resource Definitions) for all the existing resource from onPrem and create InventoryItem resources which takes some time depending upon the number of resources after PUT VMM Server operation. So it's not a bug but it take some while to populate the inventory item resource.

**Comment by neil-yechenwei** [comment]
@catriona-m , thanks for the comment. Please take another look.

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (jackofallops) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: Hi @neil-yechenwei - Whilst it's not considered a bug by the service team, asking users to put an arbitrary duration "sleep" in their code is, at best, brittle and asking for a failure somewhere down the line. We should be working with concrete states. Can we get anything back reliably from the service to know the inventory is actually ready? If nothing currently exists, can the service team make this available?

Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: I will check with service team for this concern.

Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @jackofallops , I have confirmed with service team. Below is the answer from service team. Based on the reply from service team, I think that we still need the delay expected in the flow.

```
Depending on the inventory size, it is true that it can take more than a minute time to create ARM resources for the inventory items. 
In a production level scenario, generally PUT VMM Server is a one time task and all other tasks are automated to be performed daily. During the VMM Server onboarding there is a delay expected in the flow, which we can document and ask the customers to add a delay in their automation. For the automation pipeline you can add a delay of 1 minute.

Even if we make the ask available from our end you need to make GET calls at regular interval to check for the sync status and in itself it is equivalent to a delay. I agree that this will be a better and reliable solution to the problem, but it would require service level and API changes. We will be picking this up down the line and would request you to get first release done with the current behavior and in the later release we can add this functionality.
```

**Comment by neil-yechenwei** [comment]
@jackofallops , thanks for the comment. I replied your concern. Please take another look.

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (katbyte) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @neil-yechenwei - as it stands we cannot merge this resource with a sleep like this. users should not need to worry about if the resource is created or not and be adding "time_sleep" resources to the config. when the create function exits the resource _must_ be ready to use.

thus any delays must be in the create function, and they must be dynamic and not a simple "sleep" because as steve says that is brittle and as the service team says it could be longer then 1 min. If there is no current way to determine when the create is finished/resource ready i'm not sure we can add this to the provider at this time.

Comment (hsurana06) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @catriona-m @katbyte - 
I am from the SCVMM service team and would like to clarify on the above discussion:

- If a customer wants to enable a SCVMM resource in Azure, he will need to provide the inventory Id or the UUID for the on-prem resource (which makes the unique inventory Id). 

- There are two options either to get UUID from on-prem setup or query on the inventory list API and get the inventory resources with UUID. If a customer is using ARC, will try to get the UUID from the inventory list only.

- Case 1: Now if a customer gets the UUID/InventoryId from the Inventory list, we are good to create the required Azure resource.
- Case 2: If customer gets the UUID from infra and inventory is not yet synced. We can add a polling/retry mechanism where we will check for the UUID associated inventory from the inventory list and either proceed or retry for a fixed number of time in fixed/backoff intervals before failing.

So, with this polling inventory list, we will start the resource creation if we get the inventory id in time or will fail the operation with timeout. And this way we won't need any sleep to be introduced.

Note - When the customer first onboards the VMM Server, the operator creates Custom Resource Definitions (CRDs) which are then synced to Azure via RPaaS in form of inventory items. It take some time to create all resources in Azure depending on the infra size but it is a one-time process. Subsequent inventory syncs happen only for modified CRs. And this applies to all resources like Cloud/VM Network/VM Template/VM as this is by design.

Comment (tombuildsstuff) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @hsurana06 so how can users tell when this process is completed through the API? Is there a "status" field showing the current state?

Comment (hsurana06) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: We don't have any status field to mark if complete inventory has synced or not, but if we want to check for a particular inventory Item we can query and proceed like the following example:
 
For a given resource say a cloud with uuid "A-B-C-D", if user wants to create the ScVmm.Cloud resource on Azure he needs to check if the List Inventory Items API return the Inventory with this UUID. If the list has the entry with this UUID he can proceed to create the required resource and the newly created resource will have the Status.

Comment (tombuildsstuff) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @hsurana06 are there plans to add that sync status to the API? Unfortunately we have no knowledge of what the UUIDs are going to be here, so I don't believe we can poll on the items directly?

Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @tombuildsstuff , service team confirmed that the sync would definitely be completed within ten minutes. I think we just need to set a timeout of 10 minutes and check the inventory quantity continuously every minute for 10 times. If the quantity doesn't change, then we consider the sync to be complete. So I submitted this [PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/26004) to handle this issue. Then we don't need to add the sleep in the tf config anymore.

**Comment by neil-yechenwei** [comment]
thanks for the comment. Please take another look.

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (tombuildsstuff) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @neil-yechenwei what happens if you've got a larger data set, or a slow upload? Without a sync status coming back from the API, this is essentially just a `sleep(10m)`?

Comment (sidsinghms) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @tombuildsstuff / @katbyte / @neil-yechenwei 
After deliberation within the product team, we suggest - PUT VMMServer to be not supported in terraform. We can support all other operations/resources.

PUT VMMServer (~Datacenter onboarding) is a Day 0 scenario, and from our interactions with the customers, we believe it is unlikely for them to use automation for a Day 0 scenario like this at least in the customer use cases that we've encountered. In-fact we have heard most customers specifically asking for VM CRUD support via Terraform for our service. 
For now, we have decided not to expose PUT VMMServer in terraform, while supporting all other operation. In future, when have all the right properties hydrated. we can consider adding PUT VMMServer at a later stage.

Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @sidsinghms , so you mean that TF doesn't need to support System Center Vritual Machine Manager Cloud, System Center Vritual Machine Manager Virtual Machine Template and System Center Vritual Machine Manager Virtual Network.

Comment (sidsinghms) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @neil-yechenwei 
Yup, we can deprecate/remove the VMMServer resource from TF.

The plan makes sense, looks good to us. Let's go ahead with the above-mentioned steps.

A few clarifications for us
-What are the timelines we are looking at? [To convey to our customers as well as internal tracking]
-What are the expectations from the product team going forward? [To evaluate bandwidth, we'll need from our side]

Comment (katbyte) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @neil-yechenwei / @sidsinghms - a couple things:

* for deprecation & removal it would sit around till the next major version before removal
* as such we need to atvery least fix/improve the existing resource for any users to happen to be or start to use it ignoring the deprecation
* what is the size of items uploaded and how can you be sure 10min is enough? ie what is the longest something could be taking to upload as 10m seems far to fast, i be more comfortable with 2 hours to account for slow connections and other things that could slow uploads down.
* if we do include that fix, which technically unblocks this and other SCVMM prs (as the wait will be removed from the dependant resources) should we still remove the server resource? 
  * even as a day zero step i would imagine users still would expect to use terraform to create that resource rather then have to dip into the portal, point and click, grab the resrouce id manually vs their server FQDN and logging info then slot it manually into their config. 
  * I obviously don't have all the context of your users environments but i imagine some folk might have dev/qa/prod envs & servers and would like to write 1 tf config they test before deploying into prod which would need that resoruce

so regarding removal and deprecation i'm not sure that is something we should do as its making the terraform user experiance worse and tests more opaque (magic resource ID)

thus - @neil-yechenwei could we please update that pr with the longer wait? so we can at least move that forward i would be comfortable merging it with a longer wait which would at least allow us to remove the wait from the other PRs and start to move them forward and if we do remove the server resource it wouldn't be a user facing change doer those resource and we can just update the test cases 

@sidsinghms 
- timelines for? to get the additional PRs merged? or to remove a resoruce?
- i think the only thing we would like is a commitment to is add something into the api for "syncing" or "not syncing" so we could remove the long wait from the server resource

Comment (neil-yechenwei) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown:35]: @katbyte , thanks for the suggestions. I updated that PR https://github.com/hashicorp/terraform-provider-azurerm/pull/26004. Please take a look. Thanks.

Comment (sidsinghms) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown]: @katbyte 
The idea behind deprecation/removal of the server resource was to unblock other resources to be onboarded to TF, as they are much more asked about by the customers.

Onboarding a VMMServer involves an [appliance resource](https://learn.microsoft.com/en-us/azure/azure-arc/resource-bridge/overview) being deployed in the customer environment. The parts needed before deploying servers' resource for our service to work are anyways out of scope of this effort. 
The deployment/onboarding script includes creation of a few resources including the VMMServers resource. All other resources are to be used post the [onboarding experience](https://learn.microsoft.com/en-us/azure/azure-arc/system-center-virtual-machine-manager/overview#onboard-resources-to-azure-management-at-scale) is complete.
Point being, there are anyways a few 'magic resource IDs', due to nature of our onboarding experience, so we might as well have VMMServer added in that list.

Now, since we anyways seem to be supporting VMMServers resource at least until next major version release, we are okay with your suggestion to have 2 hours wait to account for some of the edge cases. We arrived at 10 minutes, based on our P99 telemetry for VMMServers inventory sync. For the maximum inventory size we support i.e. 15K VMs, it takes up to 25 mins for the complete sync, in lab environment. If we can have a 2-hour gap. it will be great and should cover almost all our customers.

TL;DR
- If the fix unblocks other SCVMM PRs, we do need to remove the VMMServers resource from TF.
- VMMServers resource on TF, is a 'good to have' for our services. Other frequently used resources are much more needed for our customer's use cases.
   - We are okay to remove the VMMServers resource, only if it is blocking other resources. 
 
@katbyte
- We are looking for timelines for these resources to be consumed by the customers.
- Yup, well received, we'll consider adding syncing/non-syncing indicator on the resource, in the coming months.

**Comment by neil-yechenwei** [comment]
@katbyte , I noticed that the PR to add stateRefreshFunc has been merged. So I updated this PR to remove the time sleep accordingly. Below is the test result I just now triggered.

```
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/basic (540.31s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/complete (332.87s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/requiresImport (391.18s)
--- PASS: TestAccSystemCenterVirtualMachineManagerCloudSequential/scvmmCloud/update (779.31s)
```

File: website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown
Comment (katbyte) [reply website/docs/r/system_center_virtual_machine_manager_cloud.html.markdown]: @sidsinghms now that this and the other PRs are unblocked it shouldn't take more then a couple weeks for them to make their way in

**Review by katbyte — APPROVED** [review]
LGTM 🌾

**Pull request merged**
